pallets/hyperdrive/src/lib.rs

```diff
@@ -0,0 +1,257 @@
+                    .iter()
+                    .fold((vec![], vec![], vec![]), |acc, action| {
+                        let (mut added, mut updated, mut removed) = acc;
+                        match action {
+                            StateTransmitterUpdate::Add(account, activity_window) => {
+                                <StateTransmitter<T, I>>::set(
+                                    account.clone(),
+                                    activity_window.clone(),
+                                );
+                                added.push((account.clone(), activity_window.clone()))
+                            }
+                            StateTransmitterUpdate::Update(account, activity_window) => {
+                                <StateTransmitter<T, I>>::set(
+                                    account.clone(),
+                                    activity_window.clone(),
+                                );
+                                updated.push((account.clone(), activity_window.clone()))
+                            }
+                            StateTransmitterUpdate::Remove(account) => {
+                                <StateTransmitter<T, I>>::remove(account);
+                                removed.push(account.clone())
+                            }
+                        }
+                        (added, updated, removed)
+                    });
+
+            // Emit event to inform that the state transmitters were updated
+            Self::deposit_event(Event::StateTransmittersUpdate {
+                added,
+                updated,
+                removed,
+            });
+
+            Ok(())
+        }
+
+        #[pallet::call_index(1)]
+        #[pallet::weight(Weight::from_ref_time(10_000).saturating_add(T::DbWeight::get().reads_writes(1, 2)))]
+        pub fn submit_state_merkle_root(
+            origin: OriginFor<T>,
+            block: T::TargetChainBlockNumber,
+            state_merkle_root: T::TargetChainHash,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            let activity_window = <StateTransmitter<T, I>>::get(&who);
+            let current_block = <frame_system::Pallet<T>>::block_number();
+            // valid window is defined inclusive start_block, exclusive end_block
+            ensure!(
+                activity_window.start_block <= current_block
+                    && current_block < activity_window.end_block,
+                Error::<T, I>::SubmitOutsideTransmitterActivityWindow
+            );
+            ensure!(
+                block
+                    .checked_rem(&T::TransmissionRate::get())
+                    .ok_or(Error::<T, I>::CalculationOverflow)?
+                    .is_zero(),
+                Error::<T, I>::SubmitOutsideTransmissionRate
+            );
+
+            // insert merkle root proposal since all checks passed
+            // allows for constant-time validity checks
+            let accepted =
```

Comment:
The idea here was to expect at least X root submissions and then get the one with the majority of the votes. But this works for now.

Reply:
the current implementation assumes we make sure quorum size is > number of ceil(transmitters/2)